source/extensions/dynamic_modules/sdk/rust/src/lib.rs

```diff
@@ -41,3 +58,111 @@ macro_rules! declare_program_init {
+pub type ProgramInitFunction = fn() -> bool;
+
+/// The function signature for the new HTTP filter configuration function.
+///
+/// This is called when a new HTTP filter configuration is created, and it must return a new instance of the [`HttpFilterConfig`] object.
+/// Returning `None` will cause the HTTP filter configuration to be rejected.
+//
+// TODO(@mathetake): I guess there would be a way to avoid the use of dyn in the first place.
+// E.g. one idea is to accept all concrete type parameters for HttpFilterConfig and HttpFilter traits in declare_init_functions!,
+// and generate the match statement based on that.
+pub type NewHttpFilterConfigFunction = fn(
+    envoy_filter_config: EnvoyHttpFilterConfig,
+    name: &str,
+    config: &str,
+) -> Option<Box<dyn HttpFilterConfig>>;
+
+/// The global init function for HTTP filter configurations. This is set via the `declare_init_functions` macro,
+/// and is not intended to be set directly.
+pub static mut NEW_HTTP_FILTER_CONFIG_FUNCTION: NewHttpFilterConfigFunction = |_, _, _| {
+    panic!("NEW_HTTP_FILTER_CONFIG_FUNCTION is not set");
+};
+
+/// The trait that represents the configuration for an Envoy Http filter configuration.
+/// This has one to one mapping with the [`EnvoyHttpFilterConfig`] object.
+///
+/// The object is created when the corresponding Envoy Http filter config is created, and it is
+/// dropped when the corresponding Envoy Http filter config is destroyed. Therefore, the imlementation
+/// is recommended to implement the [`Drop`] trait to handle the necessary cleanup.
+pub trait HttpFilterConfig {
+    /// This is called when a HTTP filter chain is created for a new stream.
+    fn new_http_filter(&self) -> Box<dyn HttpFilter> {
+        unimplemented!() // TODO.
+    }
+}
+
+/// The trait that represents an Envoy Http filter for each stream.
+pub trait HttpFilter {} // TODO.
+
+/// An opaque object that represents the underlying Envoy Http filter config. This has one to one
+/// mapping with the Envoy Http filter config object as well as [`HttpFilterConfig`] object.
+///
+/// This is a shallow wrapper around the raw pointer to the Envoy HTTP filter config object, and it
+/// can be copied and used up until the corresponding [`HttpFilterConfig`] is dropped.
+//
+// TODO(@mathetake): make this only avaialble for non-test code, and provide a mock for testing. So that users
+// can write a unit tests for their HttpFilterConfig implementations.
+#[derive(Debug, Clone, Copy)]
+pub struct EnvoyHttpFilterConfig {
+    raw_ptr: abi::envoy_dynamic_module_type_http_filter_config_envoy_ptr,
+}
+
+#[no_mangle]
+unsafe extern "C" fn envoy_dynamic_module_on_http_filter_config_new(
+    envoy_filter_config_ptr: abi::envoy_dynamic_module_type_http_filter_config_envoy_ptr,
+    name_ptr: *const u8,
+    name_size: usize,
+    config_ptr: *const u8,
+    config_size: usize,
+) -> abi::envoy_dynamic_module_type_http_filter_config_module_ptr {
```

Comment:
In general for all unsafe entry points, do only enough unsafe work to create safe objects, then call an "inner" function to do the actual work. We will use this pattern in all entry points.

Reply:
https://github.com/envoyproxy/envoy/pull/37070/commits/f59880cb68bfa4ee2c6f77d9eb89eb17ce2187a6 done